core/templates/pages/contributor-dashboard-page/modal-templates/question-suggestion-editor-modal.controller.ts

```diff
@@ -54,7 +56,11 @@ angular.module('oppia').controller('QuestionSuggestionEditorModalController', [
     $scope.misconceptionsBySkill = {};
     $scope.misconceptionsBySkill[$scope.skill.getId()] = (
       $scope.skill.getMisconceptions());
-    ContextService.setImageSaveDestinationToLocalStorage();
+    ContextService.resetImageSaveDestination();
```

Comment:
Could you please explain (conceptually) what the old code was doing, and why this change had to be made?

Reply:
My findings on this bug is as following.

1. The initial error was thrown from [this line](https://github.com/oppia/oppia/blob/6d514f7e19cbcbf8b40601c469e5ccdc823ed05d/extensions/objects/templates/image-with-regions-editor.component.ts#L317) of `image-with-regions-editor.component.ts` while calling the `this.contextService.getExplorationId()`. Since no exploration id is set, the error was thrown. `getPreviewUrl()` was called by the `<image>` in the `image-with-regions-editor.component.html` to set href value.
2. When image saving destination is set to `ContextService.setImageSaveDestinationToLocalStorage()` images are temporarily stored in the web browser and all images are sent to the server when the question is submitted. But in this case, inorder to get the reference link to the href of `<image>`, we need to send the image to the server before the question is submitted. Hence this change was necessary in order to save images in the server before the question is submitted.

Reply:
Oh I see, thanks for explaining. You definitely need comments in the code here, at the very least. 

Some thoughts:

1. I feel like the idea of "resetting" doesn't quite make sense here, conceptually. Shouldn't we just set it?
2. Do we need to look into generalize the `this.contextService.getExplorationId()` call, if the image-with-regions-editor component isn't always going to appear in an exploration context? That seems like the better way to go IMO?
3. Can you explain why we need to "get the reference link to the href of `<image>`"? Also I think "saving images in the server before the question is submitted" is probably fine but we must make sure that this only happens after we've committed to saving the question (i.e. the user can't cancel between the images being saved and the actual question being submitted).

Also, where are images being saved for question suggestions? Is it in some temporary place that's suggestion-related and then they are copied over to the main place where question images are located, once the question is accepted? If so, then are the temporary images deleted? Just trying to understand the full workflow, so that I can advise better here.

Reply:
I thought of keeping image saving destination to local storage as it was and that would reduce the complexity of the solution. If we can get the `imageDataURI` which is defined [here](https://github.com/oppia/oppia/blob/6d514f7e19cbcbf8b40601c469e5ccdc823ed05d/extensions/objects/templates/image-editor.component.ts#L947) and return it from `getPreviewUrl ()` from [here](https://github.com/oppia/oppia/blob/6d514f7e19cbcbf8b40601c469e5ccdc823ed05d/extensions/objects/templates/image-with-regions-editor.component.ts#L317). But I am blocked how to get the `imageDataURI` from `getPreviewUrl ()`. Could you please help me on this?

Reply:
I need more info in order to help. Could you please answer my previous question in more detail?

If we have a better common understanding of how the system is working, then it will be easier to find the right solution.

Reply:
Sorry for the earlier comment without fully explaining for what was asked initially.

1. I think we do not need to reset or set the image saving destination, since there is a default destination defined i.e. `AppConstants.IMAGE_SAVE_DESTINATION_SERVER` which is exactly what is done in `ContextService.resetImageSaveDestination()`.
2. In order to eliminate this, I set a [customeContextEntity](https://github.com/oppia/oppia/blob/6b2f0288b92214499a312cba86d2b78c456a8707/core/templates/pages/contributor-dashboard-page/modal-templates/question-suggestion-editor-modal.controller.ts#L60) and the entity id set over there is taken in `getPreviewUrl()` instead of `this.contextService.getExplorationId()` -- [link](https://github.com/oppia/oppia/blob/6b2f0288b92214499a312cba86d2b78c456a8707/extensions/objects/templates/image-with-regions-editor.component.ts#L322).
3. My initial assumption was wrong here -- we still can use `imageDataURI `  which is defined [here](https://github.com/oppia/oppia/blob/6d514f7e19cbcbf8b40601c469e5ccdc823ed05d/extensions/objects/templates/image-editor.component.ts#L947) as the source for the `<image>`. But as I said in my [previous comment](https://github.com/oppia/oppia/pull/14289#discussion_r754042332), I am not able to get the `imageDataURI ` of the selected image to the image-with-regions-editor component.

Reply:
Why are we still resetting here?